Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/UpgradeData.xml

```diff
@@ -14525,11 +14612,130 @@
+        <AffectedUnitArray value="AP_MothershipPurifier"/>
+        <AffectedUnitArray value="AP_MothershipPurifierPowerSource"/>
+    </CUpgrade>
+    <CUpgrade id="AP_OptimizedEnergyUsageAiurMax">
+        <EffectArray Reference="Unit,AP_MothershipAiur,Speed" Value="0.500000"/>
+        <EffectArray Reference="Unit,AP_MothershipAiur,EnergyRegenRate" Value="1.500"/>
+        <EffectArray Operation="Set" Reference="Behavior,AP_OptimizedEnergyUsageAiur,Modification.VitalRegenArray[Energy]" Value="0"/>
+        <AffectedUnitArray value="AP_MothershipAiur"/>
+    </CUpgrade>
+    <CUpgrade id="AP_OptimizedEnergyUsageTaldarimMax">
+        <EffectArray Reference="Unit,AP_SOAMothershipv4,Speed" Value="0.500000"/>
+        <EffectArray Reference="Unit,AP_SOAMothershipv4,EnergyRegenRate" Value="1.500"/>
+        <EffectArray Reference="Unit,AP_MothershipTaldarim,Speed" Value="0.500000"/>
+        <EffectArray Reference="Unit,AP_MothershipTaldarim,EnergyRegenRate" Value="1.500"/>
+        <EffectArray Operation="Set" Reference="Behavior,AP_OptimizedEnergyUsageTaldarim,Modification.VitalRegenArray[Energy]" Value="0"/>
+        <AffectedUnitArray value="AP_SOAMothershipv4"/>
+        <AffectedUnitArray value="AP_MothershipTaldarim"/>
+    </CUpgrade>
     <CUpgrade id="AP_TriremeSolarBeam">
         <EffectArray Operation="Set" Reference="Weapon,AP_CarrierPurifierPrismaticBeam,Options[Disabled]" Value="0"/>
         <EffectArray Operation="Set" Reference="Weapon,AP_CarrierPurifierPrismaticBeamDummy,Options[Disabled]" Value="0"/>
         <AffectedUnitArray value="AP_CarrierPurifier"/>
     </CUpgrade>
+    <CUpgrade id="AP_PurifiedEmbers">
+        <EffectArray Operation="Set" Reference="Effect,AP_PurifierMothershipSecondaryBeamPersistent,FinalEffect" Value="AP_PurifiedEmbersAB"/>
+        <EditorCategories value="Race:Protoss,UpgradeType:Talents"/>
+        <AffectedUnitArray value="AP_MothershipPurifier"/>
+        <AffectedUnitArray value="AP_MothershipPurifierPowerSource"/>
+    </CUpgrade>
+    <CUpgrade id="AP_RecursivePhotons">
+        <EffectArray Operation="Set" Reference="Weapon,AP_AiurMothershipWeapon,Effect" Value="AP_AiurMothershipSlowBeamSet"/>
+        <EditorCategories value="Race:Protoss,UpgradeType:Talents"/>
+        <AffectedUnitArray value="AP_MothershipAiur"/>
+    </CUpgrade>
+    <CUpgrade id="AP_HungerForSlaughter">
+        <EffectArray Reference="Weapon,AP_SOAMothershipv4Weapon,RateMultiplier" Value="1"/>
+        <EffectArray Operation="Set" Reference="Weapon,AP_SOAMothershipv4Weapon,Effect" Value="AP_SOAMothershipv4FastBeamSet"/>
+        <EffectArray Operation="Subtract" Reference="Effect,AP_SOAMothershipv4BeamDamage,Amount" Value="19"/>
+        <EditorCategories value="Race:Protoss,UpgradeType:Talents"/>
+        <AffectedUnitArray value="AP_SOAMothershipv4"/>
+        <AffectedUnitArray value="AP_MothershipTaldarim"/>
+    </CUpgrade>

```

Comment:
An upgrade received by an item shouldn't have a drawback. This could be solved by a toggleable firing mode driven by a Behavior. Check Infested Libs that do that

Reply:
Still missing weapon toggle